### PR TITLE
Meshing for `Triangle3d` primitive

### PR DESCRIPTION
**Pull request by mweatherley**
# Objective

- Ongoing work for #10572 
- Implement the `Meshable` trait for `Triangle3d`, allowing 3d triangle primitives to produce meshes.

## Solution

The `Meshable` trait for `Triangle3d` directly produces a `Mesh`, much like that of `Triangle2d`. The mesh consists only of a single triangle (the triangle itself), and its vertex data consists of:
- Vertex positions, which are the triangle's vertices themselves (i.e. the triangle provides its own coordinates in mesh space directly)
- Normals, which are all the normal of the triangle itself
- Indices, which are directly inferred from the vertex order (note that this is slightly different than `Triangle2d` which, because of its lower dimension, has an orientation which can be corrected for so that it always faces "the right way")
- UV coordinates, which are produced as follows:
    1. The first coordinate is coincident with the `ab` direction of the triangle.
    2. The second coordinate maps to be perpendicular to the first in mesh space, so that the UV-mapping is skew-free.
    3. The UV-coordinates map to the smallest rectangle possible containing the triangle, given the preceding constraints.

Here is a visual demonstration; here, the `ab` direction of the triangle is horizontal, left to right — the point `c` moves, expanding the bounding rectangle of the triangle when it pushes past `a` or `b`:

Screenshot 2024-03-23 at 5 36 01 PM

Screenshot 2024-03-23 at 5 38 12 PM

Screenshot 2024-03-23 at 5 37 15 PM

The UV-mapping of `Triangle2d` has also been changed to use the same logic.

---

## Changelog

- Implemented `Meshable` for `Triangle3d`.
- Changed UV-mapping of `Triangle2d` to match that of `Triangle3d`.

## Migration Guide

The UV-mapping of `Triangle2d` has changed with this PR; the main difference is that the UVs are no longer dependent on the triangle's absolute coordinates, but instead follow translations of the triangle itself in its definition. If you depended on the old UV-coordinates for `Triangle2d`, then you will have to update affected areas to use the new ones which, briefly, can be described as follows:
- The first coordinate is parallel to the line between the first two vertices of the triangle.
- The second coordinate is orthogonal to this, pointing in the direction of the third point.

Generally speaking, this means that the first two points will have coordinates `[_, 0.]`, while the third coordinate will be `[_, 1.]`, with the exact values depending on the position of the third point relative to the first two. For acute triangles, the first two vertices always have UV-coordinates `[0., 0.]` and `[1., 0.]` respectively. For obtuse triangles, the third point will have coordinate `[0., 1.]` or `[1., 1.]`, with the coordinate of one of the two other points shifting to maintain proportionality.

For example: 
- The default `Triangle2d` has UV-coordinates `[0., 0.]`, `[0., 1.]`, [`0.5, 1.]`.
- The triangle with vertices `vec2(0., 0.)`, `vec2(1., 0.)`, `vec2(2., 1.)` has UV-coordinates `[0., 0.]`, `[0.5, 0.]`, `[1., 1.]`.
- The triangle with vertices `vec2(0., 0.)`, `vec2(1., 0.)`, `vec2(-2., 1.)` has UV-coordinates `[2./3., 0.]`, `[1., 0.]`, `[0., 1.]`.

### Design considerations

1. There are a number of ways to UV-map a triangle (at least two of which are fairly natural); for instance, we could instead declare the second axis to be essentially `bc` so that the vertices are always `[0., 0.]`, `[0., 1.]`, and `[1., 0.]`. I chose this method instead because it is skew-free, so that the sampling from textures has only bilinear scaling. I think this is better for cases where a relatively "uniform" texture is mapped to the triangle, but it's possible that we might want to support the other thing in the future. Thankfully, we already have the capability of easily expanding to do that with Builders if the need arises. This could also allow us to provide things like barycentric subdivision.
2. Presently, the mesh-creation code for `Triangle3d` is set up to never fail, even in the case that the triangle is degenerate. I have mixed feelings about this, but none of our other primitive meshes fail, so I decided to take the same approach. Maybe this is something that could be worth revisiting in the future across the board.

### CODE REVIEW

File: crates/bevy_render/src/mesh/primitives/dim3/triangle3d.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+};
+
+impl Meshable for Triangle3d {
+    type Output = Mesh;
+
+    fn mesh(&self) -> Self::Output {
+        let positions: Vec<_> = self.vertices.into();
+        let uvs: Vec<_> = uv_coords(self).into();
+
+        // Every vertex has the normal of the face of the triangle (or zero if the triangle is degenerate)
```
Comment (Chubercik) [inline crates/bevy_render/src/mesh/primitives/dim3/triangle3d.rs:16]:
```suggestion
        // Every vertex has the normal of the face of the triangle (or zero if the triangle is degenerate).
```

File: crates/bevy_render/src/mesh/primitives/dim2.rs [hunk 2]
```diff
@@ -197,25 +197,23 @@ impl Meshable for Triangle2d {
...
-            a / extents / 2.0 + 0.5,
-            b / extents / 2.0 + 0.5,
-            c / extents / 2.0 + 0.5,
-        ];
+        let uvs: Vec<_> = triangle3d::uv_coords(&Triangle3d::new(
+            vertices_3d[0],
+            vertices_3d[1],
+            vertices_3d[2],
+        ))
+        .into();
```
Comment (NthTensor) [inline crates/bevy_render/src/mesh/primitives/dim2.rs:210]: ✨*fancy*✨

File: crates/bevy_render/src/mesh/primitives/dim3/mod.rs [hunk 3]
```diff
@@ -4,6 +4,7 @@ mod cylinder;
 mod plane;
 mod sphere;
 mod torus;
+pub(crate) mod triangle3d;
```
Comment (NthTensor) [inline crates/bevy_render/src/mesh/primitives/dim3/mod.rs:7]: why is this in a different format to the other module declarations?

**Review by NthTensor — APPROVED** [review]
Lovely! Looks all correct to me. Really appreciate the comments on the UVs. The new UV layout seems like an unambiguous improvement too.

Do you think we could get this added to one of the primitive examples, now that it's meshed?

File: crates/bevy_render/src/mesh/primitives/dim3/mod.rs (hunk 3 above)
Comment (mweatherley) [reply crates/bevy_render/src/mesh/primitives/dim3/mod.rs:7]: It's pub(crate) so that the `Triangle2d` code can see the `Triangle3d` UV generation code

**Review by alice-i-cecile — APPROVED** [review]
Looks solid to me. I'd like to see this in an example too, but I won't block on it.

**Review by Chubercik — APPROVED** [review]
Code looks good, changes made sound reasonable.

File: crates/bevy_render/src/mesh/primitives/dim2.rs [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
     render_asset::RenderAssetUsages,
 };
 
 use super::Meshable;
 use bevy_math::{
     primitives::{
-        Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, WindingOrder,
+        Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, Triangle3d, WindingOrder,
     },
     Vec2,
```
Comment (Chubercik) [inline crates/bevy_render/src/mesh/primitives/dim2.rs:12]:
```suggestion
```

File: crates/bevy_render/src/mesh/primitives/dim2.rs [hunk 5]
```diff
@@ -1,12 +1,13 @@
...
+    mesh::primitives::dim3::triangle3d,
     mesh::{Indices, Mesh},
     render_asset::RenderAssetUsages,
 };
 
 use super::Meshable;
 use bevy_math::{
     primitives::{
-        Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, WindingOrder,
+        Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, Triangle3d, WindingOrder,
```
Comment (Chubercik) [inline crates/bevy_render/src/mesh/primitives/dim2.rs:10]:
```suggestion
        Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, Triangle3d,
        WindingOrder,
```

File: crates/bevy_render/src/mesh/primitives/dim2.rs [hunk 6]
```diff
@@ -1,14 +1,15 @@
 use crate::{
+    mesh::primitives::dim3::triangle3d,
     mesh::{Indices, Mesh},
     render_asset::RenderAssetUsages,
 };
 
 use super::Meshable;
 use bevy_math::{
     primitives::{
```
Comment (Chubercik) [inline crates/bevy_render/src/mesh/primitives/dim2.rs:9]:
```suggestion
use bevy_math::primitives::{
```

File: crates/bevy_render/src/mesh/primitives/dim2.rs [hunk 7]
```diff
@@ -1,14 +1,15 @@
...
-        Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, WindingOrder,
+        Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, Triangle3d,
+        WindingOrder,
     },
```
Comment (Chubercik) [inline crates/bevy_render/src/mesh/primitives/dim2.rs:12]:
```suggestion
    Annulus, Capsule2d, Circle, Ellipse, Rectangle, RegularPolygon, Triangle2d, Triangle3d,
    WindingOrder,
```

**Review by Chubercik — CHANGES_REQUESTED** [review]
How peculiar, my previous fmt review comment was actually a copy-paste of the CI's suggestion. Weird how that wasn't the final, desirable version.

**Pull request merged**